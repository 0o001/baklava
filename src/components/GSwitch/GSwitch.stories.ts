import { storiesOf } from '@storybook/vue';
import GSwitch from './GSwitch.vue';
import { boolean, withKnobs } from '@storybook/addon-knobs';

const stories = storiesOf('GSwitch', module);
stories.addDecorator(withKnobs);

stories
  .addParameters({ component: GSwitch })
  .add('Default', () => ({
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
    },
    components: { GSwitch },
    data () {
      return {
        checked: true,
      };
    },
    template: '<div style=\'margin:100px;\'>' +
        '<div><GSwitch v-model="checked" :disabled="disabled"/></div>' +
        '<div><GSwitch v-model="checked" :size="\'medium\'"></div>' +
        '<div><GSwitch v-model="checked" :size="\'small\'"></div>' +
        '</div>',
  }));
