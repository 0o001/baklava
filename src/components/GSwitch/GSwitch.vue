<template>
  <label :class="classNames">
    <input
      ref="checkbox"
      type="checkbox"
      @change.prevent="handleChange"
      :disabled="disabled"
    >
    <span :class="contentClassname" />
  </label>
</template>

<script lang="ts">

import $_ from '../../utils/dashUtil.ts';

export default {
  name: 'GSwitch',
  props: {
    value: {
      type: [Boolean, Object],
      default: () => false,
    },
    disabled: {
      type: Boolean,
      default: () => false,
    },
    size: {
      default: 'big',
      type: String,
    },
  },
  computed: {
    classNames () {
      return {
        'g-switch': true,
        disabled: this.disabled,
      };
    },
    contentClassname () {
      return {
        [$_(this.size)]: true,
      };
    },
  },
  methods: {
    handleChange () {
      this.$emit('input', !this.value);
      this.$emit('changeToggle');
    },
  },
  watch: {
    value () {
      this.$refs.checkbox.checked = this.value;
    },
  },
  mounted () {
    this.$refs.checkbox.checked = this.value;
  },
};
</script>

<style lang="scss" scoped>
  .g-switch {
    cursor: pointer;
    position: relative;
    display: inline-flex;
    margin: 0;

    &.disabled {
      cursor: not-allowed;
    }
    &.disabled span {
      background-color: var(--bg-grey-500);
    }
    &.disabled input:checked + span {
      background-color: var(--green-900);
    }

    input {
      display:none;
    }

    input:checked + span {
      background-color: var(--green-500);
    }

    span {
      background-color: var(--mid-grey-800);
      position: relative;
      display: inline-block;
      transition: background-color .4s ease;
    }

    span:before{
      position: absolute;
      content: "";
      top: 2px;
      left: 2px;
      background-color: white;
      -webkit-transition: .4s;
      transition: .4s;
    }

    span.-big{
      width: 76px;
      height: 40px;
      border-radius: 20px;
    }

    span.-big:before {
      height: 36px;
      width: 36px;
      border-radius: 18px;
    }

    input:checked + span.-big:before {
      transform: translateX(36px);
    }

    span.-medium{
      width: 60px;
      height: 32px;
      border-radius: 16px
    }

    span.-medium:before {
      height: 28px;
      width: 28px;
      border-radius: 14px;
    }

    input:checked + span.-medium:before {
      transform: translateX(28px);
    }

    span.-small{
      width: 44px;
      height: 24px;
      border-radius: 12px
    }

    span.-small:before {
      height: 20px;
      width: 20px;
      border-radius: 11px;
    }

    input:checked + span.-small:before {
      transform: translateX(20px);
    }
  }
</style>
