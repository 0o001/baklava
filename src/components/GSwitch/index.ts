import GSwitch from './GSwitch.vue';

export default GSwitch;
