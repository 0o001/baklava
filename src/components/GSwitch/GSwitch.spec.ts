import { shallowMount } from '@vue/test-utils';
import GSwitch from './';

describe('GSwitch', () => {
  let wrapper: any;

  const createWrapper = (options = {}) => {
    wrapper = shallowMount(GSwitch, options);
  };

  it('should match snapshot with default', () => {
    createWrapper();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should match snapshot when size small', () => {
    createWrapper({
      propsData: {
        size: 'small',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should match snapshot when handleChange fired', () => {
    createWrapper({
      propsData: {
        size: 'medium',
      },
    });

    wrapper.vm.handleChange();
    expect(wrapper.emitted().changeToggle.length).toEqual(1);
    expect(wrapper.emitted().input.length).toEqual(1);
    expect(wrapper.element).toMatchSnapshot();
  });
});
