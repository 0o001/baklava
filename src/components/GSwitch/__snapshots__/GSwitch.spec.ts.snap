// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GSwitch should match snapshot when handleChange fired 1`] = `
<label
  class="g-switch"
>
  <input
    type="checkbox"
  />
   
  <span
    class="-medium"
  />
</label>
`;

exports[`GSwitch should match snapshot when size small 1`] = `
<label
  class="g-switch"
>
  <input
    type="checkbox"
  />
   
  <span
    class="-small"
  />
</label>
`;

exports[`GSwitch should match snapshot with default 1`] = `
<label
  class="g-switch"
>
  <input
    type="checkbox"
  />
   
  <span
    class="-big"
  />
</label>
`;
